### PR TITLE
Send missing holiday reservation reminders

### PR DESCRIPTION
**Pull request by reynders**
#### Summary
- Send missing holiday reservation email reminder to those guardians or foster parents that have a daycare placed child during a holiday with a reservation deadline day after tomorrow
- On espoo prod for 27.3.2024 the getChildrenWithWithMissingReservations was Execution Time: 102.383 ms

### CODE REVIEW

File: service/src/main/kotlin/fi/espoo/evaka/emailclient/EvakaEmailMessageProvider.kt [hunk 1]
```diff
@@ -254,6 +254,24 @@ $unsubscribeEn
         )
     }
 
+    override fun missingHolidayReservationsNotification(language: Language): EmailContent {
```
Comment (Joosakur) [inline service/src/main/kotlin/fi/espoo/evaka/emailclient/EvakaEmailMessageProvider.kt:257]: Hmm, this always returns all three languages. Is it just for possible customization's sake that we even fetch the language?

File: service/src/main/kotlin/fi/espoo/evaka/emailclient/EmailClient.kt [hunk 2]
```diff
@@ -34,7 +34,10 @@ private constructor(
             traceId: String,
         ): Email? {
             val (toAddress, enabledEmailTypes) =
-                dbc.read { tx -> tx.getEmailAddressAndEnabledTypes(personId) }
+                dbc.read { tx ->
+                    tx.getEmailAddressAndEnabledTypes(personId)
+                        ?: EmailAndEnabledEmailTypes(null, null)
```
Comment (Joosakur) [inline service/src/main/kotlin/fi/espoo/evaka/emailclient/EmailClient.kt:39]: optional nitpick: this should rather log that the person was not found (which sounds unexpected and a possible bug) and return

File: service/src/main/kotlin/fi/espoo/evaka/reservations/MissingHolidayReservationsReminders.kt [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+                val personsToBeNotified =
+                    childrenWithMaybeMissingHolidayReservations
+                        .map {
+                            tx.getChildGuardiansToNotify(
+                                clock.today(),
+                                childrenWithMaybeMissingHolidayReservations.toList()
+                            )
+                        }
+                        .flatten()
+                        .toSet()
```
[the remark this answers is not in the document]
Comment (reynders) [reply service/src/main/kotlin/fi/espoo/evaka/reservations/MissingHolidayReservationsReminders.kt]: Extremely good catch!

File: service/src/main/kotlin/fi/espoo/evaka/reservations/MissingHolidayReservationsReminders.kt [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    ) {
+        val language = db.read { tx -> getLanguage(tx, msg.guardian) }
+
+        Email.create(
+                dbc = db,
+                personId = msg.guardian,
+                emailType = EmailMessageType.MISSING_HOLIDAY_ATTENDANCE_RESERVATION_NOTIFICATION,
+                fromAddress = emailEnv.sender(language),
+                content = emailMessageProvider.missingHolidayReservationsNotification(language),
+                traceId = msg.guardian.toString(),
```
[the remark this answers is not in the document]
Comment (reynders) [reply service/src/main/kotlin/fi/espoo/evaka/reservations/MissingHolidayReservationsReminders.kt:158]: After discussion it was agreed that it is ok to use guardian uuid here

File: service/src/main/kotlin/fi/espoo/evaka/emailclient/EvakaEmailMessageProvider.kt (hunk 1 above)
Comment (reynders) [reply service/src/main/kotlin/fi/espoo/evaka/emailclient/EvakaEmailMessageProvider.kt:257]: To be investigated separately

**Review by reynders — COMMENTED** [review]
mob review

**Review by Gekkio — APPROVED** [review]

**Pull request merged**